### ISSUE TITLE
Custom icon rules: Some files are not having proper icons

### ISSUE THREAD

**Issue by navaneethnpk** [FlorianWoelki/obsidian-iconize#492]
**Describe the bug**
I added few rules to automatically add icons. But some of the markdown files are having folder icons. Can anyone guide me to put proper rules? 

![image](https://github.com/FlorianWoelki/obsidian-iconize/assets/38340722/e6928fd4-29bb-49dc-bd4f-23cc7749f287)

In the page title and in the tab, the icons is coming properly. But in the side bar it coming same as folder.
![image](https://github.com/FlorianWoelki/obsidian-iconize/assets/38340722/67730e28-6ff0-4194-b2e2-29775d4625e4)

**Comment by FlorianWoelki** [comment]
Did you set any custom rules? Can you send a brief excerpt of the `data.json` file of the Iconize plugin?

**Comment by navaneethnpk** [comment]
Yes, I have set custom rules to add icons automatically to files like markdown, images and folders. 
I have attached the `data.json` file below.
```
{
  "settings": {
    "migrated": 5,
    "iconPacksPath": ".obsidian/icons",
    "fontSize": 16,
    "emojiStyle": "native",
    "iconColor": null,
    "recentlyUsedIcons": [],
    "recentlyUsedIconsSize": 5,
    "rules": [
      {
        "rule": ".md",
        "icon": "FasFileLines",
        "for": "everything",
        "order": 0
      },
      {
        "rule": ".",
        "icon": "FarFolderClosed",
        "for": "everything",
        "order": 3
      },
      {
        "rule": ".png",
        "icon": "FarImage",
        "for": "everything",
        "order": 1
      }
    ],
    "extraMargin": {
      "top": 0,
      "right": 4,
      "bottom": 0,
      "left": 0
    },
    "iconInTabsEnabled": true,
    "iconInTitleEnabled": true,
    "iconInTitlePosition": "inline",
    "iconInFrontmatterEnabled": false,
    "iconInFrontmatterFieldName": "icon",
    "iconColorInFrontmatterFieldName": "iconColor",
    "iconsBackgroundCheckEnabled": false,
    "iconsInNotesEnabled": true,
    "iconsInLinksEnabled": true,
    "iconIdentifier": ":",
    "debugMode": false
  }
}
```

**Comment by navaneethnpk** [comment]
Even though I set file icons to all of my markdown files (rule = ".md"), some of them were having folder icons (rule = "."). This was the bug I reported. While trying to understand this bug, I found three behaviors:

## 1) The order of 'Custom icon rules' in the plug-in settings is not matching with data.json
![image](https://github.com/user-attachments/assets/13858076-e1e9-4f60-9d82-6b5cd9faf7e3)
```
"rules": [
      {
        "rule": ".png",
        "icon": "FarFileImage",
        "for": "everything",
        "order": 2
      },
      {
        "rule": ".md",
        "icon": "LiFileText",
        "for": "everything",
        "order": 1
      },
      {
        "rule": ".",
        "icon": "LiFolder",
        "for": "everything",
        "order": 0
      }
    ]
```

## 2. Changing 'Custom icon rules' in the plug-in settings is temporarly fixing the issue
When I open Obsidian, the icons of some markdown files are incorrect (Screenshot-2). When I change the order of the ".md" rule, the icons become correct (Screenshot-3). However, if I close Obsidian and open it again, the icons revert to being incorrect. Additionally, if I change the order of the "." rule, the icons will also become incorrect.
Screenshot-2:
![image](https://github.com/user-attachments/assets/7d2f1abe-51e9-48a2-9a0a-b6c37c189767)
Screenshot-3:
![image](https://github.com/user-attachments/assets/7a01fb93-15f4-4afc-9a8a-5c4d654341aa)

## 3. markdown files in the subfolder of a folder is showing this wrong icon issue; files in the main folders are correct.
Screenshot-4:
![image](https://github.com/user-attachments/assets/fa548079-aef2-4ebe-abf5-1bd8334dc866)

 Hope this helps :)

**Comment by claycle** [comment]
I see this same issue. In fact, it works perfectly in one vault. I copied my rules to a new vault and I see the #3 issue (.md files with folders on them instead of file icons).

My rules are a little fancier, but not by much.

<img width="857" alt="image" src="https://github.com/user-attachments/assets/8e6b0de8-9e62-426f-9b75-eaa93aec725c">

**Comment by umbreon222** [comment]
I ran into the same problem with rules that overlap each other. For instance, I have global rules set with low priorities but a more specific rule set with the highest:
![image](https://github.com/user-attachments/assets/532405b0-acb2-438b-b87e-bf864c3fc0ba)

In this scenario, I'd expect the "01-02-2024" folder to be an alert triangle and not an alert circle. The problem occurs in a spot where the rules are applied but not awaited so it ends up being a race condition. Here's a fix for that btw (icon.ts:254):
![image](https://github.com/user-attachments/assets/a09d1239-6cd0-4f74-a96a-9bbf02e9dc8e)

There's **another scenario** where some wonky stuff is happening with re-ordering overlapping rules. When I swap the 2024 rule (specifically, when clicking arrows on the 2024 rule) with the global rule below it, only the global rule is applied. However, when using the arrows on the global rule to move priority from highest to lowest, then things work as expected. The problem here is that icons are removed from the rule whose arrows are being clicked and not also from the rule it's swapping with. When it goes to apply the rule who's now the highest priority, the icon it needs to replace gets skipped because it already has an icon. The fix for this is just to remove icons from both affected rules (customIconRule.ts:173-174):
![image](https://github.com/user-attachments/assets/74e90ae7-0348-4176-a19f-6e7f0757d81c)
 
I _wanted_ to put out PRs for both of these but I don't feel like installing pnpm to do so.
